### PR TITLE
Issue 551

### PR DESCRIPTION
**Pull request by omatrot**
This PR is about reverse engineering from an EDMX file, instead of an existing database. Why? Because The Conceptual model may have been tweaked to include only what is necessary from a business layer point of view. For instance, only specific tables/views and columns may have been selected. This still remains a database first approach.

This work has been made possible because of an old Codeplex project [brough back to .NET Core](https://github.com/omatrot/LinqToEdmx) that allows us to easily browse through the various parts of an EDMX file.
Please go to the repo above to know more about limitations.

There are  several new projects int the solution:

**ErikEJ.EntityFrameworkCore.5.Edmx** and **ErikJ.EntityFrameworkCore.Edmx** that exposes IDatabaseModelFactory. Because I could not found a factory to create an instance of IRelationalModel, I'm using IDatabaseModelFactory to return another DatabaseModel that is tailored to the conceptual part of the EDMX model. That was the whole idea of this enhancement. We don't really care about the database model, we want to concentrate on the Conceptual model.

The corresponding tests may be found in EdmxTest.cs (project NUnitTestCore).
The EDMX file that is available has been generated by connecting from VS 2012 to an instance of AdventureWorksLT2019 database.

This is preliminary work that should be discussed before moving further on.

The limitations would be documented at a later time.

### CODE REVIEW

**Comment by ErikEJ** [comment]
I will have a look.

Could you fix merge conflicts?

and maybe undo a number of unrelated project file changes?

**Comment by omatrot** [comment]
Of course

**Comment by ErikEJ** [comment]
Build error:
Error parsing solution file at C:\projects\efcorepowertools\src\GUI\EFCorePowerTools.sln: Exception has been thrown by the target of an invocation.  Error parsing the nested project section in solution file. A project with the GUID "{9212E4DF-87A4-40B3-A1E7-858C0A0F4DC3}" is listed as being nested under project "{D1EDDE24-5DC8-407E-BE37-77AD0B75759D}", but does not exist in the solution.  C:\projects\efcorepowertools\src\GUI\EFCorePowerTools.sln

**Comment by ErikEJ** [comment]
Wonder if the databasemodel that comes out of this can be used for anything?

**Comment by omatrot** [comment]
This adds no value if the database model doesn't come from the Conceptual model of the Edmx file I'm afraid.

**Comment by omatrot** [comment]
Still looking at build failures...

**Comment by ErikEJ** [comment]
C:\projects\efcorepowertools\src\GUI\ErikJ.EntityFrameworkCore.Edmx\Scaffolding\EdmxDatabaseModelFactory.cs(12,17): error CS0234: The type or namespace name 'EntityFrameworkCore' does not exist in the namespace 'Microsoft' (are you missing an assembly reference?)

**Comment by ErikEJ** [comment]
Could you remove og gitignore the .DS_Store files?

**Comment by ErikEJ** [comment]
`<PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="5.0.0" />`
Is it missing from the .5. project?

**Comment by ErikEJ** [comment]
Great, so the build passes.

What value would this bring, if a DbContext was generated based on this? 

As an initial migrations model? 

As a starting point to begin refining?

**Comment by omatrot** [comment]
The added values lies in the Conceptual model (The entities and the associated context), which would bring a ready to use DbContext. Neither migrations, nor refining necessary. The former is not necessary because we're talking "database first", the later is already available, because someone already did the work of tailoring the Store model to their needs.

**Comment by ErikEJ** [comment]
But how would you use this at runtime without mappings to a store?

**Comment by omatrot** [comment]
That is a good question, I thought I would be able to provide both.
I may be wrong, but the EF scaffolding tools seems to blindly take the store model to generate the entities configuration.
With the EDMX and your tooling, I thought I could provide just what's needed.

One of my customer, who's using EDMX (And EF6) for a long time, asked me to use EF Core in an new project with the same database. It took me so much time to remove what I wouldn't need from what has been generated from the database, that I thought it would be great to take advantage of their existing EDMX files.

**Comment by ErikEJ** [comment]
What if I merge this, and enable the command line tool to generate a DbContext from the test EDMX? 
Then you can validate the usability?

**Comment by omatrot** [comment]
Is it only a Windows tool?
If not I'd rather generate a DbContext on my Mac with your instructions, and validate it before you merge?

**Comment by ErikEJ** [comment]
No, it is a NET core console app

**Comment by omatrot** [comment]
Wat would be the process to do the generate the DbContext from the EDMX?

**Comment by ErikEJ** [comment]
That would be to modify the command line tool to understand that it should use the EDMX Databasemodel factory (for now we could use the DacpacPath parameter, and check the file extension), by modifying the ServiceBuilder in the RevEng.Core.50 project. Then run efreveng50 with F5, specifying a parameter file, that would point to the EDMX. (See the Testfiles folder in the efreveng project)

**Comment by omatrot** [comment]
Ok let me see that this evening and I'll get back to you.

**Comment by omatrot** [comment]
All right, I think that it starts to make sense to me.
May be the Conceptual part of the model could be used to generate an option file for the reverse engineering process? It seems that you could select which tables to include during generation. If we could also select columns may be this would add value. Or we could not use an option file at all and process the conceptual part during the generation?

**Comment by omatrot** [comment]
Hi @ErikEJ.
I've read the docs and seen that we could exclude columns, which is great because this is exactly what the EDMX conceptual model is about. As I said in my previous comment, I think we should discuss if you prefer we work with an option file generated from the model or process directly the conceptual model during the scaffolding.
Let me know what you think.

**Comment by ErikEJ** [comment]
I do not really understand what these conceptual model considerations are, but regardless, adding more option files than the two existing ones is not a direction I want to take.

I would like to see a model generated from an EDMX and see it used in a test app, to better understand the usefullness of this, and also to be able to consider if it is something I want to support going forward.

**Comment by omatrot** [comment]
I would say that it maps the following scaffolfing opions your tooling provide, specifically:
https://github.com/ErikEJ/EFCorePowerTools/wiki/Reverse-Engineering#choose-database-objects
https://github.com/ErikEJ/EFCorePowerTools/wiki/Reverse-Engineering#choose-options

**Comment by omatrot** [comment]
I can't provide a model generated from an EDMX without modifying the old T4 templates available in VS 2012, and it will lack all the entities configuration code. This is why I would like to make progress with your tooling.

**Comment by ErikEJ** [comment]
What's preventing you from doing that now, at least from the efreveng commabd line tool?

**Comment by omatrot** [comment]
I would say an option file example that specifies all the options documented in the links above. I've looked in https://github.com/ErikEJ/EFCorePowerTools/tree/master/src/GUI/efreveng/TestFiles. I may be wrong but I can't see how to specify columns to include for a specific table.

**Comment by ErikEJ** [comment]
It is ExcludedColumns ! Best option is to run the tool against a database to generate a config file, and exclude some columns to get the json format.

**Comment by ErikEJ** [comment]
Let me know if there is anything I can assist with.

**Comment by omatrot** [comment]
Sure.
I'll install VS 2019 and the tooling to run it against the database I used to produce the EDMX.

**Comment by ErikEJ** [comment]
??? You do not need to install VS in order to run the reveng.exe command line tool! It is just a .NET Core console app.

**Comment by omatrot** [comment]
I need to generate a config file from the VS extension I presume.

**Comment by ErikEJ** [comment]
Of course, forgot about that.

File: src/GUI/ErikEJ.EntityFrameworkCore.5.Edmx/ErikEJ.EntityFrameworkCore.5.Edmx.csproj [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    <Folder Include="Scaffolding\" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="..\ErikEJ.EntityFrameworkCore.Edmx\Scaffolding\EdmxDatabaseModelFactory.cs">
+      <Link>Scaffolding\EdmxDatabaseModelFactory.cs</Link>
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="linq2edmx" Version="0.0.1-pre" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="5.0.1" />
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.5.Edmx/ErikEJ.EntityFrameworkCore.5.Edmx.csproj:21]: 5.0.0 ?

**Review by ErikEJ — CHANGES_REQUESTED** [review]
Looks good - I can enable in UI later. Please revert the unrelated .csproj change re target platform and remove unneeded inline code comments, and I will merge and integrate.

Was the generated Dbcontext of any use?

**Comment by omatrot** [comment]
I'm happy to say that EDMX scaffolding is working. It is not perfect yet because it is limited to SQL Server backends. I'm used to PostgreSQL as well so It will be the next backend to support.
You'll find a new test file 'edmx.config.json' generated with the VS Extension, alongside a zip file that is the VS extension scaffolding result.
You could launch the revenge.exe command line tool against this file provided that you've fixed the various paths that are not Windows compatible. When it is done you could compare the output files with the ones generated by your tool.

I've modified ServiceProviderBuilder.cs like the following:

```
            // Add database provider services
            switch (options.DatabaseType)
            {
                case DatabaseType.Undefined:
                    var edmxProvider = new SqlServerDesignTimeServices();
                    edmxProvider.ConfigureDesignTimeServices(serviceCollection);

                    serviceCollection.AddSingleton<IDatabaseModelFactory, EdmxDatabaseModelFactory>();

                    break;
```

I think we should discuss how to integrate this the right way because, and EDMX could point to all ADO.NET relational databases.

What I've seen so far:
- You have no summary on entities columns because the information is missing from the EDMX.
- I've explicitly unchecked a column in the Customer Entity but it is still there (bug on my side I presume).
- The RowGUID columns are generated as Byte[] instead of GUID
- The entities configuration is limited by the informations available in the EDMX (no index, default values...)

To summarize, and clarify things I hope:
- The EF DatabaseModel metadata maps to the EDMX Storage Model.
- Your config files and/or VS Extension scaffolding options maps to the EDMX Conceptual Model.
  - This is why It seems interesting to me to leverage the conceptual model directly from the reveng.exe command line tool instead of using a config file.
  - Another option would be to generate a config file from the EDMX as well.
 - The main idea is to leverage Database First efforts from the early 2010s, not to generate a dababase or even migrations from the generated code.

I will now go on to run the code against the database.

**Comment by ErikEJ** [comment]
>> The main idea is to leverage Database First efforts from the early 2010s, not to generate a dababase or even migrations from the generated code.

Yes, makes perfectly sense, but is the generated context usable at runtime against a database, if not, it seems quite useless to me.

File: src/GUI/ErikEJ.EntityFrameworkCore.5.Edmx/ErikEJ.EntityFrameworkCore.5.Edmx.csproj (hunk 1 above)
Comment (omatrot) [reply src/GUI/ErikEJ.EntityFrameworkCore.5.Edmx/ErikEJ.EntityFrameworkCore.5.Edmx.csproj]: Sorry I was in the process of writing a big comment.
I removed the Microsoft.EntityFrameworkCore.Relational reference

**Comment by omatrot** [comment]
I'm writing the code right now to run against the database (running in a container on Mac) from VS 2019 Windows running in Parallels Desktops...

File: src/GUI/ErikEJ.EntityFrameworkCore.5.Edmx/ErikEJ.EntityFrameworkCore.5.Edmx.csproj (hunk 1 above)
Comment (omatrot) [reply src/GUI/ErikEJ.EntityFrameworkCore.5.Edmx/ErikEJ.EntityFrameworkCore.5.Edmx.csproj]: And now fixed it to 5.0.0.
Is that what you'd expect?
Could you clarify "unneeded inline code comments"?

**Comment by ErikEJ** [comment]
You enjoy pain? :smile:

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    // Because the EDMX format was not designed with the database creation in mind, the SSDL part of the schema won't contains default values, extended properties, primary keys, unique constraints and index informations.
+    public class EdmxDatabaseModelFactory : IDatabaseModelFactory
+    {
+        // SQL Server db types
+        private static readonly ISet<string> SQLServerDateTimePrecisionTypes = new HashSet<string> { "datetimeoffset", "datetime2", "time" };
+
+        private static readonly ISet<string> SQLServerMaxLengthRequiredTypes
+            = new HashSet<string> { "binary", "varbinary", "char", "varchar", "nchar", "nvarchar" };
+
+        // FIXME Add PostgreSQL db types
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:30]: Remove comment

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+
+        // FIXME Add MySql db types
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:32]: Remove comment

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+            var schemas = options.Schemas;
+            var tables = options.Tables;
+
+            var dbModel = new DatabaseModel
+            {
+                DatabaseName = Path.GetFileNameWithoutExtension(edmxPath),
+                DefaultSchema = schemas.Count() > 0 ? schemas.First() : "dbo"
+            };
+
+            // Detect the EDMX file version
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:94]: Only support EDMX 3, so remove comment

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+            // FIXME Provide an alternative to not load the whole document in memory /!\
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:95]: Not important - remove comment

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+            XDocument edmxFile = XDocument.Load(edmxPath);
+
+            var edmxVersion = ((XElement)edmxFile.FirstNode).FirstAttribute.Value;
+
+            // FIXME Assume EDMX version is 3.0 for now
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:100]: Remove comment

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+                    GetColumnsV3(item, dbTable/*, typeAliases, model.GetObjects<TSqlDefaultConstraint>(DacQueryScopes.UserDefined).ToList()*/, edmxv3);
+                    GetPrimaryKeyV3(item, dbTable);
+
+                    dbModel.Tables.Add(dbTable);
+                }
+
+                foreach (var item in items)
+                {
+                    GetForeignKeysV3(edmxv3, item, dbModel);
+                    // Unique constraints are not available in the model
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:124]: Remove commented out code

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+
+        public DatabaseModel Create(DbConnection connection, DatabaseModelFactoryOptions options)
+        {
+            throw new NotImplementedException();
+        }
+
+        private void GetColumnsV3(LinqToEdmx.Model.StorageV3.EntityTypeStore item, DatabaseTable dbTable/*, IReadOnlyDictionary<string, (string storeType, string typeName)> typeAliases, List<TSqlDefaultConstraint> defaultConstraints,*/, LinqToEdmx.EdmxV3 model)
+        {
+            var tableColumns = item.Properties;
+                //.Where(i => !i.GetProperty<bool>(Column.IsHidden)
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:143]: Remove dead code

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+                //&& i.ColumnType != ColumnType.ColumnSet
+                //// Computed columns not supported for now
+                //// Probably not possible: https://stackoverflow.com/questions/27259640/get-datatype-of-computed-column-from-dacpac
+                //&& i.ColumnType != ColumnType.ComputedColumn
+                //);
+
+            foreach (var col in tableColumns)
+            {
+                // Default constraints are not available in the model.
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:152]: Remove comments

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+
+                var dbColumn = new DatabaseColumn
+                {
+                    Table = dbTable,
+                    Name = col.Name,
+                    IsNullable = col.Nullable,
+                    StoreType = storeType,
+                    DefaultValueSql = defaultValue,
+
+                    // ComputedColumnSQL is not available in the model
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:165]: Remove dead code

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+                    //    : storeType == "rowversion"
+                    //        ? ValueGenerated.OnAddOrUpdate
+                    //        : default(ValueGenerated?)
+                };
+                if (storeType == "rowversion")
+                {
+                    dbColumn["ConcurrencyToken"] = true;
+                }
+
+                //var description = model.GetObjects<TSqlExtendedProperty>(DacQueryScopes.UserDefined)
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:180]: Remove dead code

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+                    return $"{dataTypeName}(max)";
+                }
+
+                return $"{dataTypeName}({col.MaxLength})";
+            }
+
+            // Integral type
+            return col.Type.ToString();
+
+            // FIXME PostgreSQL
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:233]: Remove comment

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 13]
```diff
@@ -0,0 +1,379 @@
...
+
+            // FIXME MySQL
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:235]: Remove comment

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 14]
```diff
@@ -0,0 +1,379 @@
...
+            var pk = table.Key;
+            var primaryKey = new DatabasePrimaryKey
+            {
+                // We do not have information about the primary key name in the model.
+                // So we're making it up
+                Name = string.Concat("PK_", dbTable.Name),
+                Table = dbTable
+            };
+
+            // We do not know if the primary key is clustered because we're not limited to SQL Server Here.
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:251]: Not important, so remove

File: src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+                    // The table that contains the foreign key constraint (Dependent in the Edmx)
+                    Table = table,
+                    // The table to which the columns are constrained (Principal in the Edmx)
+                    PrincipalTable = principalTable,
+                    
+                };
+
+                var end = associationSet.Ends[0];
+
+                // OnDelete = ConvertToReferentialAction(fk.DeleteAction);
```
Comment (ErikEJ) [inline src/GUI/ErikEJ.EntityFrameworkCore.Edmx/Scaffolding/EdmxDatabaseModelFactory.cs:325]: Remove

File: src/GUI/SqlCeToolbox.DDEX4/SqlCeToolbox.DDEX4.csproj [hunk 16]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
```
Comment (ErikEJ) [inline src/GUI/SqlCeToolbox.DDEX4/SqlCeToolbox.DDEX4.csproj:5]: Revert

File: src/GUI/UnitTests/UnitTests.csproj [hunk 17]
```diff
@@ -1,9 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
```
Comment (ErikEJ) [inline src/GUI/UnitTests/UnitTests.csproj:6]: Revert

File: src/GUI/efpt30.core/efpt30.core.csproj [hunk 18]
```diff
@@ -6,6 +6,7 @@
     <AssemblyName>efpt</AssemblyName>
     <RootNamespace>Modelling</RootNamespace>
     <NoWarn>$(NoWarn);EF1001</NoWarn>
+    <Configurations>Release;Debug</Configurations>
```
Comment (ErikEJ) [inline src/GUI/efpt30.core/efpt30.core.csproj:9]: Revert

File: src/GUI/efpt50.core/efpt50.core.csproj [hunk 19]
```diff
@@ -6,6 +6,7 @@
     <AssemblyName>efpt</AssemblyName>
     <RootNamespace>Modelling</RootNamespace>
     <NoWarn>$(NoWarn);EF1001</NoWarn>
+    <Configurations>Release;Debug</Configurations>
```
Comment (ErikEJ) [inline src/GUI/efpt50.core/efpt50.core.csproj:9]: Revert

**Review by ErikEJ — CHANGES_REQUESTED** [review]

**Comment by omatrot** [comment]
The generated context is almost working against the database.
I have this problem with these columns (SSDL):

`[rowguid] [uniqueidentifier] ROWGUIDCOL  NOT NULL,`

That are scaffolded as .NET Byte[] instead of Guid.

In the Factory code, I'm detecting the store type as 'uniqueidentifier' and thus returning 'rowversion' for DatabaseColumn.StoreType. Does that ring any bell for you?

**Comment by ErikEJ** [comment]
Let me know when I can review again !

**Comment by omatrot** [comment]
That should be good for comments and reverts

**Comment by ErikEJ** [comment]
Feel free to resolve comments!

rowversion = timestamp = byte[]

unique identifier = Guid

**Comment by omatrot** [comment]
Capture d’écran 2021-02-21 à 17 40 49

Success !

**Comment by omatrot** [comment]
> Feel free to resolve comments!

Do I still need to do something about this one?

**Comment by ErikEJ** [comment]
Fantastic! 

I will go through the PR again soon.

**Comment by ErikEJ** [comment]
I will resolve as I review again.

**Comment by omatrot** [comment]
Thanks, this will be all for today for me. Have a good evening.

File: src/GUI/UnitTests/UnitTests.csproj [hunk 20]
```diff
@@ -1,8 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
```
Comment (ErikEJ) [inline src/GUI/UnitTests/UnitTests.csproj:6]: Remove this line

File: src/GUI/SqlCeToolbox.DDEX4/SqlCeToolbox.DDEX4.csproj [hunk 21]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
     <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
```
Comment (ErikEJ) [inline src/GUI/SqlCeToolbox.DDEX4/SqlCeToolbox.DDEX4.csproj:6]: Remove this line

File: src/GUI/UnitTests/UnitTests.csproj [hunk 22]
```diff
@@ -1,8 +1,9 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (ErikEJ) [inline src/GUI/UnitTests/UnitTests.csproj:2]: Revert this change

**Review by ErikEJ — CHANGES_REQUESTED** [review]

**Comment by ErikEJ** [comment]
LGTM, only few unneeded changes

**Comment by omatrot** [comment]
> LGTM, only few unneeded changes

Sorry about that.
VS 2019 for Mac and Windows are not completely aligned about projects options I'm afraid...
